sympy/core/equation.py

```diff
@@ -0,0 +1,359 @@
+        """
+        return self._applyfunc(func, *args, **kwargs, Eqn_apply_side='both')
+    
+    def applylhs(self, func, *args, **kwargs):
+        """
+        If lhs side of the equation has a defined subfunction (attribute) of name ``func``, that will be applied
+        instead of the global function. The operation is applied to only the lhs.
+        """
+        return self._applyfunc(func, *args, **kwargs, Eqn_apply_side='lhs')
+
+    def applyrhs(self, func, *args, **kwargs):
+        """
+        If rhs side of the equation has a defined subfunction (attribute) of name ``func``, that will be applied
+        instead of the global function. The operation is applied to only the rhs.
+        """
+        return self._applyfunc(func, *args, **kwargs, Eqn_apply_side='rhs')
+
+#####
+# Overrides of binary math operations
+#####
+
+    _op_priority = 11.0 # This makes sure the rules for equations are applied before those for expressions
+                        # which have _op_priority = 10.0
+    @classmethod
+    def _binary_op(cls, a, b, opfunc_ab):
+        if isinstance(a, Equation) and not isinstance(b, Equation):
+            return Equation(opfunc_ab(a.lhs, b), opfunc_ab(a.rhs, b))
+        elif isinstance(b, Equation) and not isinstance(a, Equation):
+            return Equation(opfunc_ab(a, b.lhs), opfunc_ab(a, b.rhs))
+        elif isinstance(a, Equation) and isinstance(b, Equation):
+            return Equation(opfunc_ab(a.lhs, b.lhs), opfunc_ab(a.rhs, b.rhs))
+        else:
+            raise TypeError('One of a or b should be an equation')
+
+    def __add__(self, other):
+        return self._binary_op(self, other, lambda a, b: a + b)
+
+    def __radd__(self, other):
+        return self._binary_op(other, self, lambda a, b: a + b)
+
+    def __mul__(self, other):
+        return self._binary_op(self, other, lambda a, b: a * b)
+
+    def __rmul__(self, other):
+        return self._binary_op(other, self, lambda a, b: a * b)
+
+    def __sub__(self, other):
+        return self._binary_op(self, other, lambda a, b: a - b)
+
+    def __rsub__(self, other):
+        return self._binary_op(other, self, lambda a, b: a - b)
+
+    def __truediv__(self, other):
+        return self._binary_op(self, other, lambda a, b: a / b)
+
+    def __rtruediv__(self, other):
+        return self._binary_op(other, self, lambda a, b: a / b)
+
+    def __mod__(self, other):
+        return self._binary_op(self, other, lambda a, b: a % b)
+
+    def __rmod__(self,other):
+        return self._binary_op(other, self, lambda a, b: a % b)
+
+    def __pow__(self, other):
+        return self._binary_op(self, other, lambda a, b: a ** b)
+
+    def __rpow__(self, other):
+        return self._binary_op(other, self, lambda a, b: a ** b)
+
+    def _eval_power(self, other):
+        return self.__pow__(other)
+    
+#####
+# Output helper functions
+#####
+    def __repr__(self):
+        return(str(self.lhs)+self.relop+str(self.rhs))
```

Comment:
We should define `_print_Equation` in the printer code rather than defining `__repr__` or `__str__` here.

See e.g.:
https://github.com/sympy/sympy/blob/master/sympy/printing/str.py#L715

Reply:
I am open to suggestions here. I have found the printing code extraordinarily opaque. I am completely unclear how to extend it properly.

Reply:
got working through `sympy.printing`.